src/timestepper.cpp

```diff
@@ -173,7 +180,7 @@ Vec TimeStepper::solveODE(int initid, Vec rho_t0, int n0){
 }
 
 
-void TimeStepper::solveAdjointODE(int initid, Vec rho_t0_bar, Vec finalstate, double Jbar_penalty, double Jbar_penalty_dpdm, double Jbar_energy_penalty) {
+Vec TimeStepper::solveAdjointODE(int initid, Vec rho_t0_bar, Vec finalstate, double Jbar_penalty, double Jbar_penalty_dpdm, double Jbar_energy_penalty, int n0) {
```

Comment:
Please explain why you need to return 'x'

Reply:
This `Vec` is the gradient with respect to initial condition. No extra operation is needed for this, simply returning it from `solveAdjointODE`. We need this gradient except for time=0, but `TimeStepper::solveAdjointODE` will not know whether the time window is the first one or not. Determining this and adding the intermediate condition gradient are handled in `OptimProblem::evalGradF`, see line 958-963.

Reply:
Please add a comment about this at the top of the function

Reply:
The comment is already added at header file `src/timestepper.hpp` line 69.